ergo-lib/src/wallet/multi_sig.rs

```diff
@@ -0,0 +1,914 @@
+        let mut real_proposition: Vec<SigmaBoolean> = vec![SigmaBoolean::ProofOfKnowledge(
+            SigmaProofOfKnowledgeTree::ProveDlog(pk1.clone()),
+        )];
+        let simulated_proposition: Vec<SigmaBoolean> = Vec::new();
+        let mut bag_c = bag_for_multi_sig(
+            cand.clone(),
+            &real_proposition,
+            &simulated_proposition,
+            &proof,
+        )
+        .unwrap();
+        bag_c.add_hint(Hint::CommitmentHint(CommitmentHint::OwnCommitment(
+            carol_secret_commitment.first().unwrap().clone(),
+        )));
+        bag_c.add_hint(Hint::CommitmentHint(CommitmentHint::RealCommitment(
+            bag2.first().unwrap().clone(),
+        )));
+        let proof3 = prover3
+            .prove(
+                &tree_and,
+                &Env::empty(),
+                ctx.clone(),
+                message.as_slice(),
+                &bag_c,
+            )
+            .unwrap();
+        proof = Vec::from(proof3.proof.clone());
+        let proof_byte3: ProofBytes = proof3.proof;
+        real_proposition = vec![
+            SigmaBoolean::ProofOfKnowledge(SigmaProofOfKnowledgeTree::ProveDlog(pk1)),
+            SigmaBoolean::ProofOfKnowledge(SigmaProofOfKnowledgeTree::ProveDlog(pk3)),
+        ];
+        let mut bag_b =
+            bag_for_multi_sig(cand, &real_proposition, &simulated_proposition, &proof).unwrap();
+        bag_b.add_hint(Hint::CommitmentHint(CommitmentHint::OwnCommitment(
+            bob_secret_commitment.first().unwrap().clone(),
+        )));
+        let proof2 = prover2
+            .prove(
+                &tree_and,
+                &Env::empty(),
+                ctx.clone(),
+                message.as_slice(),
+                &bag_b,
+            )
+            .unwrap();
+        let proof_byte2: ProofBytes = proof2.proof;
+        let verifier = TestVerifier;
+
+        assert!(
+            !verifier
+                .verify(
+                    &tree_and,
+                    &Env::empty(),
+                    ctx.clone(),
+                    proof_byte1,
+                    message.as_slice()
+                )
+                .unwrap()
+                .result,
+            "{}",
+            "{}"
+        );
+
+        assert!(
+            !verifier
+                .verify(
+                    &tree_and,
+                    &Env::empty(),
+                    ctx.clone(),
+                    proof_byte3,
+                    message.as_slice()
+                )
+                .unwrap()
+                .result,
+            "{}",
+            "{}"
+        );
+
+        assert!(
+            verifier
+                .verify(
+                    &tree_and,
+                    &Env::empty(),
+                    ctx,
+                    proof_byte2,
+                    message.as_slice()
+                )
+                .unwrap()
+                .result,
+            "{}",
+            "{}"
+        );
+    }
```

Comment:
Although many scenarios from the Scala spec are using `AtLeast` conjecture which we don't have implemented yet there is a scenario that we're missing here - 
https://github.com/ScorexFoundation/sigmastate-interpreter/blob/795c176debfe105646a7deef3f01bd537b79c81d/sigmastate/src/test/scala/sigmastate/utxo/DistributedSigSpecification.scala#L391-L431 or-and-or with both dlog and dht.

Reply:
I added new test in new commits